Reduce exp() argument modulo 2*I*pi

gschintgen opened the pull request:
#### References to other Issues or PRs
Fixes #17216

#### Brief description of what is fixed or changed
`exp()` arguments with an imaginary part that is a rational multiple of pi
will now be reduced mod `2*pi*I`. (In fact the principal value will be retained.)

Before:
```
In [8]: exp(9*I*pi/4) - exp(I*pi/4)
Out[8]: -exp(I*pi/4) + exp(9*I*pi/4)

In [9]: simplify(_)
Out[9]: -(-1)**(1/4) + exp(9*I*pi/4)
```

After:
```
In [4]: exp(9*I*pi/4) - exp(I*pi/4)
Out[4]: 0
```


#### Other comments


#### Release Notes


<!-- BEGIN RELEASE NOTES -->

* functions
  * `exp()` now reduces its argument mod `2*pi*I`.
<!-- END RELEASE NOTES -->

gschintgen committed:
Reduce exp() argument modulo 2*pi*I

exp() arguments with an imaginary part that is a rational multiple of pi
will now be reduced mod 2*pi*I.

Before:

In [8]: exp(9*I*pi/4) - exp(I*pi/4)
Out[8]: -exp(I*pi/4) + exp(9*I*pi/4)

In [9]: simplify(_)
Out[9]: -(-1)**(1/4) + exp(9*I*pi/4)

After:

In [4]: exp(9*I*pi/4) - exp(I*pi/4)
Out[4]: 0

gschintgen committed:
Add tests for exp() periodicity handling

oscarbenjamin commented:
I wonder if something like this should evaluate as well:
```julia
In [12]: n = Symbol('n', integer=True)                                                                   

In [13]: exp(2*n*pi*I)                                                                                   
Out[13]: 
 2⋅ⅈ⋅π⋅n
ℯ       

In [14]: simplify(_)                                                                                     
Out[14]: 1

In [15]: exp(n*pi*I)                                                                                     
Out[15]: 
 ⅈ⋅π⋅n
ℯ     

In [16]: simplify(_)                                                                                     
Out[16]: 
    n
(-1) 
```
It seems sin and cos both auto-evaluate for `n*pi`.

gschintgen commented:
> 
> 
> I wonder if something like this should evaluate as well:
> ...
> 
> It seems sin and cos both auto-evaluate for `n*pi`.

Right, that should not be too difficult to add. I'll look into it. Thanks!
(And of course I'll fix the failing doctest that I forgot to test locally.)

gschintgen commented:
> 
> 
> I wonder if something like this should evaluate as well:
> 
> ```julia
> In [12]: n = Symbol('n', integer=True)                                                                   
> 
> In [13]: exp(2*n*pi*I)                                                                                   
> Out[13]: 
>  2⋅ⅈ⋅π⋅n
> ℯ       
> 
> In [14]: simplify(_)                                                                                     
> Out[14]: 1
> 
> In [15]: exp(n*pi*I)                                                                                     
> Out[15]: 
>  ⅈ⋅π⋅n
> ℯ     
> 
> In [16]: simplify(_)                                                                                     
> Out[16]: 
>     n
> (-1) 
> ```
> 
> It seems sin and cos both auto-evaluate for `n*pi`.

It seems there's already *some* code for it, at least for the first case you mentioned:
```
In [47]: n = Symbol('n', integer=True)

In [48]: exp(n*2*pi*I)
Out[48]: exp(2*I*pi*n)

In [49]: _.refine()
Out[49]: 1

In [50]: exp(n*pi*I)
Out[50]: exp(I*pi*n)

In [51]: _.refine()
Out[51]: exp(I*pi*n)
```
This is based on the new/future assumptions system:
https://github.com/sympy/sympy/blob/8ca4a683d58ac1f61cfd2e4dacf7f58b9c0fefab/sympy/functions/elementary/exponential.py#L206-L224

So if I were to implement automatic evaluation it would (partially) supersede the code that's already there:
- At instantiation of an exp object, the code in `eval` will directly simplify the expression, making the code in `refine` seemingly redundant;
- **but** the `refine` code must be left in, because its philosophy is that some time after instantiation the user may decide to change the current assumptions, e.g. make `n` an integer, or directly pass additional assumptions to `refine` in order to have the expression (re)evaluated/simplified.

Is my understanding correct so far?

So, what's the best/recommended practice here? Is it possible/encouraged/discouraged to call `refine` from within `eval`? I wouldn't want to add a commit to the PR that would only make a mess of it...
Or should I just ignore the new assumption system, consider it as "demo" code, and add the relevant simplifications to `eval()` based on the `is_...` properties?

I guess the correct answer for now would be the latter option, but any clarification would be appreciated.

gschintgen commented:
Oops, please disregard my previous comment: I've just realized that the exact same code from `refine` is also part of `eval`!
Only it doesn't seem to work. Shouldn't `exp(2*n*pi*I)` evaluate to `1` given the code that is already part of `eval`? (with `n` declared as integer of course.)

gschintgen commented:
Sorry for spamming, I got it now: in `eval` there's a `.is_number` check first. That excludes symbolic integers... I'll add a (hopefully) proper commit tomorrow.

smichr commented on sympy/functions/elementary/exponential.py:
```diff
@@ -262,6 +262,12 @@ def eval(cls, arg):
                             return -S.ImaginaryUnit
                         elif ask(Q.odd(coeff + S.Half)):
                             return S.ImaginaryUnit
+                    elif coeff.is_Rational:
```
```suggestion
                    elif coeff.is_rational:
```

This will allow symbolic coefficients to pass.

gschintgen replied:
Hi, thanks for looking through the changes! Wouldn't this proposed change entail that in case of a symbolic argument to exp(), another term (multiple of `2*pi*I`) would be added to the argument and hence it'd grow slightly longer and arguably more complex? Whereas for `is_Rational` the modulo-reduction would only be applied if it's an actual (albeit minor) simplification.
I may well be missing something though.

smichr replied:
I think what you have now will catch this case in the `(2*coeff).is_integer` block.

gschintgen committed:
Adapt a doctest in physics/matrices.py to new exp() simplifications

gschintgen commented:
@oscarbenjamin 
Concerning the simplification of `exp(2*pi*n*I)` etc. I found the following:  

I'd have to remove the check for `if arg.is_number or arg.is_Symbol` before isolating the coefficient of `pi*I`. This is working just fine and works great. Except that performance has become abysmal. (E.g. `test("_exponential")` taking 8 seconds instead of 2.

I found the culprit to be the new assumption system. Below I check an expression that's actually passed (directly or indirectly) as argument to exp somewhere in "test_exponential.py":
```
In [2]: coeff = (3*I*pi*floor(-arg(x)/pi + S.Half)).as_coefficient(pi*I)

In [3]: coeff.is_even

In [4]: ask(Q.even(coeff))
```
If you run the above commands you'll see that `.is_even` returns immediately, while ask introduces a considerable delay. (Minimal example: `ask(Q.even(arg(x)/pi))`)

As far as I can see there are two options:
- implement the additional exp simplification by taking out the `.is_number` restriction and replace the new assumptions by the old ones; or
- leave those simplifications out for now.

As of now I've only added a commit to fix the doctest failure.

oscarbenjamin commented:
Generally the new assumptions system isn't used anywhere because it's too slow and not ready (#17144).

Where are the new assumptions used in this case?

gschintgen commented:
> 
> 
> Generally the new assumptions system isn't used anywhere because it's too slow and not ready (#17144).
> 
> Where are the new assumptions used in this case?

It's already used in `exp.eval()`:
https://github.com/sympy/sympy/blob/b5a349d9c3b0e58bf24737d06d7890d89e89f296/sympy/functions/elementary/exponential.py#L252-L264

(BTW: is it even possible to have an `arg` that `is_Mul` and `is_Symbol`?)

The code quoted above is used in current master for `exp(4*I*pi)` and the like. It's that part that I referred to in my previous question.

oscarbenjamin commented:
That seems to have been added prematurely in #9766 when code from refine (new assumptions) was moved to eval (old assumptions). Eventually something like that will probably want to happen all over the codebase but I don't think it's agreed how it should be done. For example it probably makes sense to add an `assumptions` argument to eval that can be passed to ask. Without that there's no point in using the new assumptions here.

For now I would revert that to the old assumptions e.g. `if (2*coeff).is_integer`.

gschintgen committed:
Enable exp() simplification for symbolic argument

- Simplify exp() automatically if it results in 1, -1, I or -I (also
symbolically).
- Add more tests for exp() simplification.

smichr commented on sympy/functions/elementary/exponential.py:
```diff
@@ -250,18 +250,23 @@ def eval(cls, arg):
-                            return -S.ImaginaryUnit
-                        elif ask(Q.odd(coeff + S.Half)):
-                            return S.ImaginaryUnit
+            coeff = arg.as_coefficient(S.Pi*S.ImaginaryUnit)
```
Since you are working with a Mul, `.as_coefficient` -- vs `.coeff` -- is probably ok.

smichr commented:
This looks good to me.
